server/project.go

```diff
@@ -571,19 +554,72 @@ func (s *Server) registerProjectRoutes(g *echo.Group) {
 	})
 }
 
+func (s *Server) createVCSWebhook(ctx context.Context, vcsType vcsPlugin.Type, webhookEndpointID, secretToken, accessToken, instanceURL, externalRepoID string) (webhookID string, err error) {
+	// Create a new webhook and retrieve the created webhook ID
+	var webhookCreatePayload []byte
+	switch vcsType {
+	case vcsPlugin.GitLabSelfHost:
+		webhookCreate := gitlab.WebhookCreate{
+			URL:                   fmt.Sprintf("%s/%s/%s", s.profile.ExternalURL, gitlabWebhookPath, webhookEndpointID),
+			SecretToken:           secretToken,
+			PushEvents:            true,
+			EnableSSLVerification: false, // TODO(tianzhou): This is set to false, be lax to not enable_ssl_verification
+		}
+		webhookCreatePayload, err = json.Marshal(webhookCreate)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to marshal request body for creating webhook")
+		}
+	case vcsPlugin.GitHubCom:
+		webhookPost := github.WebhookCreateOrUpdate{
+			Config: github.WebhookConfig{
+				URL:         fmt.Sprintf("%s/%s/%s", s.profile.ExternalURL, githubWebhookPath, webhookEndpointID),
+				ContentType: "json",
+				Secret:      secretToken,
+				InsecureSSL: 1, // TODO: Allow user to specify this value through api.RepositoryCreate
+			},
+			Events: []string{"push"},
+		}
+		webhookCreatePayload, err = json.Marshal(webhookPost)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to marshal request body for creating webhook")
+		}
+	}
+	webhookID, err = vcsPlugin.Get(vcsType, vcsPlugin.ProviderConfig{}).CreateWebhook(
+		ctx,
+		common.OauthContext{
+			AccessToken: accessToken,
+			// We use refreshTokenNoop() because the repository isn't created yet.
+			Refresher: refreshTokenNoop(),
+		},
+		instanceURL,
+		externalRepoID,
+		webhookCreatePayload,
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create webhook")
+	}
+	return webhookID, nil
+}
+
 // refreshToken is a token refresher that stores the latest access token configuration to repository.
-func (s *Server) refreshToken(ctx context.Context, repositoryID int) common.TokenRefresher {
+func (s *Server) refreshToken(ctx context.Context, webURL string) common.TokenRefresher {
 	return func(token, refreshToken string, expiresTs int64) error {
-		if _, err := s.store.PatchRepository(ctx, &api.RepositoryPatch{
-			ID:           repositoryID,
-			UpdaterID:    api.SystemBotID,
-			AccessToken:  &token,
-			ExpiresTs:    &expiresTs,
-			RefreshToken: &refreshToken,
-		}); err != nil {
+		repos, err := s.store.FindRepository(ctx, &api.RepositoryFind{
+			WebURL: &webURL,
+		})
+		if err != nil {
 			return err
 		}
-		return nil
+		for _, repo := range repos {
+			_, err = s.store.PatchRepository(ctx, &api.RepositoryPatch{
```

Comment:
I am thinking to introduce a PATCH by web_url, so we don't need to iterate here and guarantee consistency

Reply:
[I used to want to do this.](https://github.com/bytebase/bytebase/pull/2667/commits/41d97360481d7576d84de8394250012df871ebe7)

Reply:
The current PATCH is WHERE id = $%d

What I am suggesting it to UPDATE xxx WHERE web_url = yyy

Reply:
Yeah, I didn't do it later because I ran into this problem

Reply:
Done. Make a small change to the original patch instead of introducing a new one. PTAL.